rust_snuba/ARCHITECTURE.md

```diff
@@ -0,0 +1,101 @@
+# rust-snuba consumers
+
+Snuba has two ways to ingest data into clickhouse: The old, Python-based consumers, and this Rust consumer.
+
+This document is about the Rust one, but due to it starting out as a
+line-by-line port of the Python consumer, the higher-level architecture also
+applies to Python.
+
+## Entrypoint
+
+`rust_snuba` is a library crate that is being compiled with PyO3/maturin into a
+Python wheel. This Python wheel exposes a `consumer` function which is used
+from within the Snuba (Python) CLI to launch `snuba rust-consumer`.
+Configuration parsing, CLI parsing and runtime config are still handled in
+Python so as to avoid unnecessary code duplication and potentially multiple
+sources of truth for config.
+
+## Steps
+
+`rust_snuba` then uses the Rust port of Arroyo to define its own consumer. The following steps (processing strategies) exist (see `factory.rs`):
+
+1. `HealthCheck` step, touches a file in irregular intervals to signal to k8s
+   that the consumer is still alive.
+2. `SetJoinTimeout` step, resets the join timeout to 0 for the next steps. This
+   marks a section where it's ok to drop in-flight progress during rebalancing.
+3. Message transformers, basically a threadpool to parse messages, validate
+   them and turn them into rows. This can call both back into Python code for
+   when the MessageProcessor hasn't been ported to Rust yet, but most datasets
+   have a Rust-native processor.
+4. `Reduce` opens batches by opening a connection to clickhouse, and writes
+   rows into them one-by-one. There is only one batch at a time, and the size
+   of a batch is constrained by `--max-batch-time-ms` and `--max-batch-size`.
+5. `SetJoinTimeout` is reset to "no timeout" here. Between this and the
+   previous section, everything will be dropped by Arroyo during rebalancing as
+   parsing and validating messages has no side effects. Even sending data to
+   ClickHouse is idempotent for as long as the batch is not closed.
+5. `ClickhouseWriterStep` is responsible for closing batches. It takes a batch
```

Comment:
Am I also understanding correctly that the underlying Reduce state is responsible for signaling to the ClickhouseWriterStep that a batch is closed?

Reply:
arroyo in general: the reduce step decides when a batch is done, and then forwards it to next step.

in snuba: reduce step decides when the batch should be closed, and forwards it to clickhouse writer step. clickhouse writer step is the one to close the connection to clickhouse.

so it depends on what you mean by closing the batch. reduce decides when a batch is not supposed to get more rows, but from clickhouse POV it doesn't do the closing, that's for the next step.